source/pape-manual-pages/tcpclient.1
.TH tcpclient 1
.SH NAME
tcpclient \- creates an outgoing TCP connection.
.SH SYNOPSIS
.B tcpclient
[
.I opts
]
.I host
.I port
.I prog
.SH DESCRIPTION
.I opts
is a series of getopt-style options.
.I host
is one argument.
.I port
is one argument.
.I prog
consists of one or more arguments. 

.B tcpclient
attempts to connect to a TCP server. If it is successful, it runs
.IR prog ,
with descriptor 6 reading from the network and descriptor 7 writing to the
network. It also sets up several environment variables (see
.BR tcp-environ (5)
).

The server's address is given by
.I host
and
.IR port .
.I port
may be a name from /etc/services or a number.
.I host
may be 0, referring to the local machine, or a dotted-decimal IP address, or
a host name; it is fed through qualification using dns_ip4_qualify.

If the server has several IP addresses,
.B tcpclient
tries each address in turn.
.SH OPTIONS
.B General options:
.TP
.B \-q
Quiet. Do not print error messages. 
.TP
.B \-Q
(Default.) Print error messages. 
.TP
.B \-v
Verbose. Print error messages and status messages. 
.P
.B Connection options:
.TP
.B \-T \fIx\fR+\fIy
Give up on the connection attempt after
.I x\fR+\fIy
seconds. Default: 2+58. When a host has several IP addresses,
.B tcpclient
tries to connect to the first IP address, waits
.I x
seconds, tries to connect to the second IP address, waits
.I x
seconds, etc.; then it retries each address that timed out, waiting
.I y
seconds per address. You may omit
.RI + y
to skip the second try. Before version 0.88,
.BR tcpclient (1)
will use only
.I x
(default: 60). 
.TP
.B \-i \fIlocalip
Use
.I localip
as the IP address for the local side of the connection; quit if
.I localip
is not available. Normally
.B tcpclient
lets the operating system choose an address. 
.TP
.B \-p \fIlocalport
Use
.I localport
as the TCP port for the local side of the connection; quit if
.I localport
is not available. Normally
.B tcpclient
lets the operating system choose a port. 
.TP
.B \-d
Delay sending data for a fraction of a second whenever the remote host is
responding slowly. This is currently the default, but it may not be in the
future; if you want it, set it explicitly. 
.TP
.B \-D
Never delay sending data; enable TCP_NODELAY. 
.P
.B Data-gathering options:
.TP
.B \-h
(Default.) Look up the remote host name in DNS to set the environment variable
$TCPREMOTEHOST. 
.TP
.B \-H
Do not look up the remote host name in DNS; remove the environment variable
$TCPREMOTEHOST. 
.TP
.B \-l \fIlocalname
Do not look up the local host name in DNS; use
.I localname
for the environment variable $TCPLOCALHOST. A common choice for localname is
0.
.TP
.B \-r
(Default.) Attempt to obtain $TCPREMOTEINFO from the remote host.
.TP
.B \-R
Do not attempt to obtain $TCPREMOTEINFO from the remote host. 
.TP
.B \-t \fIn
Give up on the $TCPREMOTEINFO connection attempt after
.I n
seconds. Default: 26. 
.SH SEE ALSO
tcpserver(1),
tcprules(1),
tcprulescheck(1),
argv0(1),
fixcrio(1),
recordio(1),
rblsmtpd(1),
who@(1),
date@(1),
finger@(1),
http@(1),
tcpcat(1),
mconnect(1),
tcp-environ(5)

http://cr.yp.to/ucspi-tcp.html


source/tcpclient.c
#include <sys/types.h>
#include <sys/param.h>
#include <netdb.h>
#include <unistd.h>
#include "sig.h"
#include "exit.h"
#include "sgetopt.h"
#include "uint16.h"
#include "fmt.h"
#include "scan.h"
#include "str.h"
#include "ip4.h"
#include "uint16.h"
#include "socket.h"
#include "fd.h"
#include "stralloc.h"
#include "buffer.h"
#include "error.h"
#include "strerr.h"
#include "pathexec.h"
#include "timeoutconn.h"
#include "remoteinfo.h"
#include "dns.h"

#define FATAL "tcpclient: fatal: "
#define CONNECT "tcpclient: unable to connect to "

void nomem(void)
{
  strerr_die2x(111,FATAL,"out of memory");
}
void usage(void)
{
  strerr_die1x(100,"tcpclient: usage: tcpclient \
[ -hHrRdDqQv ] \
[ -i localip ] \
[ -p localport ] \
[ -T timeoutconn ] \
[ -l localname ] \
[ -t timeoutinfo ] \
host port program");
}

int verbosity = 1;
int flagdelay = 1;
int flagremoteinfo = 1;
int flagremotehost = 1;
unsigned long itimeout = 26;
unsigned long ctimeout[2] = { 2, 58 };

char iplocal[4] = { 0,0,0,0 };
uint16 portlocal = 0;
char *forcelocal = 0;

char ipremote[4];
uint16 portremote;

char *hostname;
static stralloc addresses;
static stralloc moreaddresses;

static stralloc tmp;
static stralloc fqdn;
char strnum[FMT_ULONG];
char ipstr[IP4_FMT];

char seed[128];

static
void
optarg_num ( unsigned long * u )
{
  unsigned int pos ;
  pos = scan_ulong(optarg,u); 
  if (0 == pos || optarg[pos])
    strerr_die3x(111,FATAL, optarg, " is not a number");
}

int
main(int argc,char **argv)
{
  unsigned long u;
  int opt;
  char *x;
  int j;
  int s;
  int cloop;

  dns_random_init(seed);

  close(6);
  close(7);
  sig_ignore(sig_pipe);
 
  while ((opt = getopt(argc,argv,"dDvqQhHrRi:p:t:T:l:")) != opteof)
    switch(opt) {
      case 'd': flagdelay = 1; break;
      case 'D': flagdelay = 0; break;
      case 'v': verbosity = 2; break;
      case 'q': verbosity = 0; break;
      case 'Q': verbosity = 1; break;
      case 'l': forcelocal = optarg; break;
      case 'H': flagremotehost = 0; break;
      case 'h': flagremotehost = 1; break;
      case 'R': flagremoteinfo = 0; break;
      case 'r': flagremoteinfo = 1; break;
      case 't': optarg_num(&itimeout); break;
      case 'T': j = scan_ulong(optarg,&ctimeout[0]);
		if (optarg[j] == '+') {
		  unsigned int k ;
		  ++j;
		  k = scan_ulong(optarg + j,&ctimeout[1]);
		  if (0 == k || optarg[j + k]) strerr_die3x(111,FATAL, optarg + j, " is not a number");
		} else 
		  if (0 == j || optarg[j]) strerr_die3x(111,FATAL, optarg, " is not a number");
		break;
      case 'i': j = ip4_scan(optarg,iplocal); 
		if (0 == j || optarg[j]) strerr_die3x(111,FATAL, optarg, " is not an IP address");
		break;
      case 'p': optarg_num(&u); portlocal = u; break;
      default: usage();
    }
  argv += optind;

  if (!verbosity)
    buffer_2->fd = -1;

  hostname = *argv;
  if (!hostname) usage();
  if (str_equal(hostname,"")) hostname = "127.0.0.1";
  if (str_equal(hostname,"0")) hostname = "127.0.0.1";

  x = *++argv;
  if (!x) usage();
  if (!x[scan_ulong(x,&u)])
    portremote = u;
  else {
    struct servent *se;
    se = getservbyname(x,"tcp");
    if (!se)
      strerr_die3x(111,FATAL,"unable to figure out port number for ",x);
    portremote = ntohs(se->s_port);
    /* i continue to be amazed at the stupidity of the s_port interface */
  }

  if (!*++argv) usage();

  if (!stralloc_copys(&tmp,hostname)) nomem();
  if (dns_ip4_qualify(&addresses,&fqdn,&tmp) == -1)
    strerr_die4sys(111,FATAL,"temporarily unable to figure out IP address for ",hostname,": ");
  if (addresses.len < 4)
    strerr_die3x(111,FATAL,"no IP address for ",hostname);

  if (addresses.len == 4) {
    ctimeout[0] += ctimeout[1];
    ctimeout[1] = 0;
  }

  for (cloop = 0;cloop < 2;++cloop) {
    if (!stralloc_copys(&moreaddresses,"")) nomem();
    for (j = 0;j + 4 <= addresses.len;j += 4) {
      s = socket_tcp();
      if (s == -1)
        strerr_die2sys(111,FATAL,"unable to create socket: ");
      if (socket_bind4(s,iplocal,portlocal) == -1)
        strerr_die2sys(111,FATAL,"unable to bind socket: ");
      if (timeoutconn(s,addresses.s + j,portremote,ctimeout[cloop]) == 0)
        goto CONNECTED;
      close(s);
      if (!cloop && ctimeout[1] && (errno == error_timeout)) {
	if (!stralloc_catb(&moreaddresses,addresses.s + j,4)) nomem();
      }
      else {
        strnum[fmt_ulong(strnum,portremote)] = 0;
        ipstr[ip4_fmt(ipstr,addresses.s + j)] = 0;
        strerr_warn5(CONNECT,ipstr," port ",strnum,": ",&strerr_sys);
      }
    }
    if (!stralloc_copy(&addresses,&moreaddresses)) nomem();
  }

  _exit(111);



  CONNECTED:

  if (!flagdelay)
    socket_tcpnodelay(s); /* if it fails, bummer */

  if (!pathexec_env("PROTO","TCP")) nomem();

  if (socket_local4(s,iplocal,&portlocal) == -1)
    strerr_die2sys(111,FATAL,"unable to get local address: ");

  strnum[fmt_ulong(strnum,portlocal)] = 0;
  if (!pathexec_env("TCPLOCALPORT",strnum)) nomem();
  ipstr[ip4_fmt(ipstr,iplocal)] = 0;
  if (!pathexec_env("TCPLOCALIP",ipstr)) nomem();

  x = forcelocal;
  if (!x)
    if (dns_name4(&tmp,iplocal) == 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  if (!pathexec_env("TCPLOCALHOST",x)) nomem();

  if (socket_remote4(s,ipremote,&portremote) == -1)
    strerr_die2sys(111,FATAL,"unable to get remote address: ");

  strnum[fmt_ulong(strnum,portremote)] = 0;
  if (!pathexec_env("TCPREMOTEPORT",strnum)) nomem();
  ipstr[ip4_fmt(ipstr,ipremote)] = 0;
  if (!pathexec_env("TCPREMOTEIP",ipstr)) nomem();
  if (verbosity >= 2)
    strerr_warn4("tcpclient: connected to ",ipstr," port ",strnum,0);

  x = 0;
  if (flagremotehost)
    if (dns_name4(&tmp,ipremote) == 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  if (!pathexec_env("TCPREMOTEHOST",x)) nomem();

  x = 0;
  if (flagremoteinfo)
    if (remoteinfo(&tmp,ipremote,portremote,iplocal,portlocal,itimeout) == 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  if (!pathexec_env("TCPREMOTEINFO",x)) nomem();

  if (fd_move(6,s) == -1)
    strerr_die2sys(111,FATAL,"unable to set up descriptor 6: ");
  if (fd_copy(7,6) == -1)
    strerr_die2sys(111,FATAL,"unable to set up descriptor 7: ");
  sig_uncatch(sig_pipe);
 
  pathexec(argv);
  strerr_die4sys(111,FATAL,"unable to run ",*argv,": ");
}
